doc: CAN: Extend CAN doc and move it to networking

alexanderwachter opened the pull request:
This commit adds a lot more documentation to the Controller Area
Netwok API. The documentation is moved to the networking section.

Commit:
doc: CAN: Extend CAN doc and moved it to networking

This commit adds a lot more documentation to the Controller Area
Netwok API. The documentation is move to the networking section.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented:
ping @karstenkoenig

jukkar commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,254 @@
+
+Sending
+*******
+
```
Should we also have some information about SocketCAN, at least mention that we support that and pointing a link to sample app.

alexanderwachter replied:
Sure. I'll add a link to SocketCAN in the overview.

jukkar commented on doc/reference/networking/can.rst:
```diff
@@ -0,0 +1,9 @@
+.. _net_canbus:
```
Typo in commit msg s/is move/is moved/

jukkar commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,273 @@
+The send wire from the controller to the transceiver is called CAN TX, and the
+receive wire is called CAN RX. These wires use the logic levels whereas the bus
+level is interpreted differential between CAN H and CAN L.
+Both lines, CAN H and CAN L, at the same level, is called the recessive state.
```
This sentence sounds weird, not sure what we should say here.

alexanderwachter replied:
Rewritten

jukkar commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,273 @@
+The termination avoids reflections on the bus and keeps CAN L and CAN H at the
+same level in idle state.
+The dominant state always overrides the recessive state.
+This structure is called a wired-AND.
```
I wonder if a small picture would help to understand the thing here. The explanation above sounds complicated if one is not a CAN expert.

alexanderwachter replied:
I have pictures from my masters' thesis. Not sure if I can use them here. I would somehow need to add a copyright to not run into problems regarding plagiarism when submitting the thesis.

alexanderwachter replied:
I could add them in a few months when the thesis was accepted :-)

alexanderwachter replied:
Used my LaTex tikz skills to drow some nice pictures :-)

jukkar replied:
I noticed the .svg files, and started to wonder that if you did the pictures using LaTex, would it be possible to have the originals also here? That way if the images need changes, it is possible to re-use the original files instead of starting from scratch.

dbkinder replied:
If you haven't found the draw.io web site, that's a dandy tool for editing and saving SVG images in a form that can be edited (again) using draw.io, without creating another file.

alexanderwachter replied:
I think we should not add the LaTex sources. These are six files that are not compiled into the doc.

jukkar commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,273 @@
+state for a specific amount of time. Therefore you have to make sure that CAN RX
+is high, at least for a short time. This is also necessary for loopback mode.
+
+CAN uses Identifiers to identify the frame instead of addresses to identify a
```
Should the `Identifiers` be written as `identifiers`?

alexanderwachter replied:
Maybe "so-called Identifiers"

jukkar commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,273 @@
+
+CAN uses Identifiers to identify the frame instead of addresses to identify a
+node. This identifier can either have 11-bit width (Standard or Basic Frame) or
+29-bit in case of an Extended Frame. The zephyr API supports both Standard and
```
Probably s/zephyr/Zephyr/

karstenkoenig commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,280 @@
+    LOG_ERR("Unable to attach isr [%d]", filter_id);
+  }
+
+The snipped above shows how to attach a filter with an Interrupt callback.
```
typo 'snippet' and 'interrupt' in lower case

karstenkoenig commented:
Building the docs throws a lot of doxygen (I think) warnings on can.h that I presume is due to the use of 'filter' in both @param and @return 
Could probably adjust the the `filter id` in the return to just `filter_id` to resolve this

`/home/karsten/zephyrproject/zephyr/include/drivers/can.h:426: warning: argument 'filter' from the argument list of can_attach_isr has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:404: warning: argument 'filter' from the argument list of can_attach_msgq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:380: warning: argument 'filter' from the argument list of can_attach_workq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:426: warning: argument 'filter' from the argument list of can_attach_isr has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:404: warning: argument 'filter' from the argument list of can_attach_msgq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:380: warning: argument 'filter' from the argument list of can_attach_workq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:380: warning: argument 'filter' from the argument list of can_attach_workq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:404: warning: argument 'filter' from the argument list of can_attach_msgq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:426: warning: argument 'filter' from the argument list of can_attach_isr has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:380: warning: argument 'filter' from the argument list of can_attach_workq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:404: warning: argument 'filter' from the argument list of can_attach_msgq has muliple @param documentation sections
/home/karsten/zephyrproject/zephyr/include/drivers/can.h:426: warning: argument 'filter' from the argument list of can_attach_isr has muliple @param documentation sections
`

karstenkoenig commented:
Overall good high level introduction to CAN and then moving on to the API and how to use it with examples!

alexanderwachter commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,280 @@
+
+This basic sample sends a CAN frame with standard identifier 0x123 and eight
+bytes of data. When passing NULL as the callback, as shown in this example,
+the send function blocks until the frame is sent and acknowledged by at least
```
There is an ACK slot at the end of the frame (just before the EOF sequence)

karstenkoenig replied:
Oh wow, I didn't know, good thing I read your documentation now I guess :-)

alexanderwachter commented:
@karstenkoenig Thanks a lot for the feedback. Hopefully addressed all your concerns.
Most of the issues/questions regarding CAN and Zephyr were problems with the hardware.
People didn't connect a transceiver, for example. That's why I also focused on a general CAN explanation.

Commit:
include: drivers: CAN: changed return parameter description

Changed return parameter description from
"@retval filter id on success"  to
"@retval filter_id on success".
This change suppresses the doxy warnings.

Signed-off-by: Alexander Wachter <alexander.wachter@student.tugraz.at>

alexanderwachter commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,280 @@
+   CAN RX is high, at least for a short time. This is also necessary for
+   loopback mode.
+
+.. image:: can_transceiver.svg
```
Good catch! The pictures with the nodes had only two nodes originally.
I added a third one because it looks better :-)

karstenkoenig reviewed (approved):
LGTM, thanks!

henrikbrixandersen commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+arbitration phase. During the arbitration phase, write collisions are allowed.
+They resolve by the fact that dominant bits override recessive bits.
+Nodes monitor the bus and notice their transmission being overridden
+abort their transmission. It results in the fact that lower number identifiers
```
`and abort their transmission.`?

alexanderwachter replied:
Yes. They abort. They may, depending on the config, try again when the bus is in idle state again.

henrikbrixandersen commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+work-queue provided. In this case, it is the system work queue. Blocking is
+generally allowed in the callback but could result in a frame backlog when it is
+not limited. For the reason of a backlog, a ring-buffer is applied for every
+attached filter. The size of this buffer can is in Kconfig. This function is not
```
`The size of this buffer can be set in Kconfig.`?

alexanderwachter replied:
Yes.
CAN_WORKQ_FRAMES_BUF_CNT

alexanderwachter replied:
Oh, you mean the garbage i wrote :-)

henrikbrixandersen replied:
Yes ;-)

alexanderwachter commented:
@henrikbrixandersen We can either merge it now and someone sends a followup PR, or we have to wait until i get my hands on my PC again (on 28th of September).

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+This method is called masking.
+As an example, a mask with 11 bits set for standard or 29 bits set for extended
+identifiers must match perfectly.
+Bits that are set to zero in the mask don't care for the matching.
```
How about:
```
Bits that are set to zero in the mask are ignored when matching an identifier.
```

alexanderwachter replied:
Done.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+As an example, a mask with 11 bits set for standard or 29 bits set for extended
+identifiers must match perfectly.
+Bits that are set to zero in the mask don't care for the matching.
+The filters are implemented in hardware for the most CAN controllers.
```
Let's merge these sentences:
```
Most CAN controllers implement a limited number of filters in hardware.
The number of filters is also limited in Kconfig to save memory.
```

alexanderwachter replied:
Done.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+Additionally, the number of filters is limited in Kconfig, to save memory.
+
+
+`CAN Wikipedia article <https://en.wikipedia.org/wiki/CAN_bus>`_.
```
Let's put some words around this link:
```
You can read more about CAN bus in this
`CAN Wikipedia article <https://en.wikipedia.org/wiki/CAN_bus>`__.
```

alexanderwachter replied:
Done.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+          LOG_ERR("Sending failed [%d]", ret);
+  }
+
+This basic sample sends a CAN frame with standard identifier 0x123 and eight
```
I'd recommend putting the description before the sample code (instead of after).

alexanderwachter replied:
Moved all descriptions before the samples.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+          return can_send(can_dev, &frame, K_FOREVER, tx_irq_callback, "Sender 1");
+  }
+
+This example shows how to send a frame with extended identifier 0x1234567 and
```
Same here. I'd recommend putting the description before the sample code (instead of after).

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+    LOG_ERR("Unable to attach isr [%d]", filter_id);
+  }
+
+This example shows how to attach a callback from a work-queue.
```
Is "This example" the code above or the code below?  I'd recommend having the description before the code.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+    ... do something with the frame ...
+  }
+
+Here an example for `can_attach_msgq` is shown. With this function, it is
```
"Here"?  Above or below?

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+Samples
+*******
+
+You can find a ready to build sample for the Zephyr CAN API here:
```
Better would be:
```
We have two ready-to-build samples demonstrating use of the Zephyr CAN API: 
:ref:`CAN sample <can-sample>` and 
:ref:`SocketCAN sample <socket-can-sample>`.
```

alexanderwachter replied:
Done.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+terminate the bus. The dominant state always overrides the recessive state.
+This structure is called a wired-AND.
+
+.. warning::
```
Is this really a "warning"?  I'd recommend just a normal paragraph.

alexanderwachter replied:
Yes. This is the number one pitfall.

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+.. warning::
+
+   CAN controllers can only initialize when the bus is in the idle (recessive)
+   state for a specific amount of time. Therefore you have to make sure that
```
How long is a "specific amount of time" and "for a short time"?  Are these defined in the CAN spec?

alexanderwachter replied:
I found some notes in the data-sheets. Need to check the standard for that,

alexanderwachter replied:
It is 11 bit. Updated the docu accordingly

alexanderwachter commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,282 @@
+arbitration phase. During the arbitration phase, write collisions are allowed.
+They resolve by the fact that dominant bits override recessive bits.
+Nodes monitor the bus and notice their transmission being overridden
+abort their transmission. It results in the fact that lower number identifiers
```
Done.

alexanderwachter commented:
I added two more paragraphs. One for the bit timing and the other for mus states.

karstenkoenig commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,335 @@
+Error-frames can either be error passive or error active, depending on the state
+of the controller.
+In case the controller is in error active state, it sends six consecutive
+dominant bits, which his is a violation of the stuffing rule that all nodes can
```
which his is  -> 'his' is wrong

dbkinder replied:
yep, I suspect "his" should be deleted

alexanderwachter replied:
Done. Thanks for the review!

dbkinder commented on doc/reference/networking/can_api.rst:
```diff
@@ -0,0 +1,335 @@
+*******
+
+We have two ready-to-build samples demonstrating use of the Zephyr CAN API
+:ref:`Zephyr CAN sample <can-sample>`,
```
since there are only two items, replace the comma with the word "and"

alexanderwachter replied:
Done. Thanks for reviewing it again.

dbkinder reviewed (changes requested):
a few minor edits, otherwise, LGTM!

dbkinder reviewed (approved):
doc changes LGTM, thanks.
